esphome/components/bme680_bsec/bme680_bsec.cpp

```diff
@@ -35,27 +47,29 @@ void BME680BSECComponent::setup() {
     return;
   }
 
-  if (this->sample_rate_ == SAMPLE_RATE_ULP) {
-    const uint8_t bsec_config[] = {
-#include "config/generic_33v_300s_28d/bsec_iaq.txt"
-    };
-    this->set_config_(bsec_config);
-  } else {
-    const uint8_t bsec_config[] = {
-#include "config/generic_33v_3s_28d/bsec_iaq.txt"
-    };
-    this->set_config_(bsec_config);
-  }
-  this->update_subscription_();
-  if (this->bsec_status_ != BSEC_OK) {
+  // Initialize the BSEC library
+  if (this->reinit_bsec_lib_() != 0) {
     this->mark_failed();
     return;
   }
 
+  // Load the BSEC library state from storage
   this->load_state_();
 }
 
-void BME680BSECComponent::set_config_(const uint8_t *config) {
+void BME680BSECComponent::set_config_() {
+  const uint8_t bsec_config_ulp[] = {
+#include "config/generic_33v_300s_28d/bsec_iaq.txt"
+  };
+  const uint8_t bsec_config_lp[] = {
+#include "config/generic_33v_3s_28d/bsec_iaq.txt"
+  };
+  const uint8_t *config = NULL;
+  if (this->sample_rate_ == SAMPLE_RATE_ULP) {
+    config = bsec_config_ulp;
+  } else {
+    config = bsec_config_lp;
+  }
   uint8_t work_buffer[BSEC_MAX_WORKBUFFER_SIZE];
   this->bsec_status_ = bsec_set_configuration(config, BSEC_MAX_PROPERTY_BLOB_SIZE, work_buffer, sizeof(work_buffer));
```

Comment:
Probably worth having some logging on error here:
```suggestion
  this->bsec_status_ = bsec_set_configuration(config, BSEC_MAX_PROPERTY_BLOB_SIZE, work_buffer, sizeof(work_buffer));
  if (this->bsec_status_ != BSEC_OK) {
    ESP_LOGW(TAG, "Failed to set configuration (BSEC Error Code %d)", this->bsec_status_);
  }
```

In hindsight I'm not sure why these are all warnings and not ESP_LOGE but it's probably best to be consistent for now. I might PR switching them all over later.

Reply:
OK. However, I assume this error won't appear in case of stack overflow (so it's unrelated to the bootloop problem). Am I wrong?

Reply:
> > Hmm... The `Loaded BSEC library state` is the very last thing that `BME680BSECComponent::setup()` does, so the crash is outside of the bme680_bsec code. May it be a stack exhaustion problem? On my ESP32 I have no problems, and I have no ESP8266 :( But that increases my suspects about a stack exhaustion problem. Maybe you can try with the original code, but only adding the `bsec_state_data*` variables in bme680_bsec.h ? If the problem persists, that may be the root cause!
> 
> Seems to be the config arrays that push things over the edge. If I restore the previous behaviour whereby the config is switched between and passed into `set_config_` I can get up and running. For reasons I'm not sure I understand, doing effectively the same in an if inside `set_config_` does not appear to work.

I thought about this issue some time. And actually, it makes sense.
Explanation: the config arrays are 454 bytes, and `BSEC_MAX_WORKBUFFER_SIZE=2048`.

With the old code:

```
--->[stack usage: +0 bytes]
if (this->sample_rate_ == SAMPLE_RATE_ULP) {
--->[stack usage: +0 bytes]
  const uint8_t bsec_config[] = {
#include "config/generic_33v_300s_28d/bsec_iaq.txt"
  };
--->[stack usage: +454 bytes]
  this->set_config_(bsec_config);
--->[stack usage: +2502 bytes]
} else {
--->[stack usage: +0 bytes]
  const uint8_t bsec_config[] = {
#include "config/generic_33v_3s_28d/bsec_iaq.txt"
  };
--->[stack usage: +454 bytes]
  this->set_config_(bsec_config);
--->[stack usage: +2502 bytes]
}
--->[stack usage: +0 bytes]
```

With the new code:

```
--->[stack usage: +0 bytes]
  const uint8_t bsec_config_ulp[] = {
#include "config/generic_33v_300s_28d/bsec_iaq.txt"
  };
--->[stack usage: +454 bytes]
  const uint8_t bsec_config_lp[] = {
#include "config/generic_33v_3s_28d/bsec_iaq.txt"
  };
--->[stack usage: +908 bytes]
  const uint8_t *config = NULL;
  if (this->sample_rate_ == SAMPLE_RATE_ULP) {
    config = bsec_config_ulp;
  } else {
    config = bsec_config_lp;
  }
  uint8_t work_buffer[BSEC_MAX_WORKBUFFER_SIZE];
--->[stack usage: +2956 bytes]
```

This is only a logic thinking, and it may not be the actual point-to-point stack usage as the compiler might do its magic, but I think the total counts might be as I calculated (only accounting for the big buffers). This would even be worse when I'll add the other configurations for 1,8V supply voltage, so I think it's best to go back to the old behavior.

Also, while researching I read [this](https://arduino-esp8266.readthedocs.io/en/latest/faq/a02-my-esp-crashes.html#other-causes-for-crashes), and particularly in the _Stack_ section they say:

_Large arrays on the stack, such as uint8_t buffer[2048] should be avoided on the stack and should be dynamically allocated instead (consider smart pointers)._

So I think I'll make some experiment with smart pointers in this function. We may extended that to every work_buffer in a future PR. Will commit new code as soon as I get some more time.

Thanks for your testing!

Reply:
> OK. However, I assume this error won't appear in case of stack overflow (so it's unrelated to the bootloop problem). Am I wrong?

You are correct that the logging isn't directly related. I'd just broken something whilst testing/debugging and it was useful to have the error status output.

Interested to see what you come up with for the buffers and config load. Neither C++ nor embedded are really my forte!

Reply:
I've just been having a bit of a play around. You might be interested in https://github.com/trvrnrth/esphome/commit/857e2c77ce266086f802f9de6024298d00791539

It seemed like allocating a work buffer once and then sharing it for all instances made the most sense. Alternatively it could be made per-instance at the expense of more heap. Or on demand with a bit more effort:
```
std::unique_ptr<uint8_t[]> work_buffer{new uint8_t[BSEC_MAX_WORKBUFFER_SIZE]};
this->bsec_status_ = bsec_set_configuration(config.data(), BSEC_MAX_PROPERTY_BLOB_SIZE, reinterpret_cast<uint8_t *>(work_buffer.get()), BSEC_MAX_WORKBUFFER_SIZE);
```

Interested to hear your thoughts on approach.

I also switched to loading the config into vectors so that data ends up on the heap as well. As they're not massive I wonder if it would be faster with multiple sensors to allocate them them once per instance at setup time rather than on every re-init, but haven't benchmarked that as I only have a single sensor on my test rig.

Reply:
> I've just been having a bit of a play around. You might be interested in [trvrnrth@857e2c7](https://github.com/trvrnrth/esphome/commit/857e2c77ce266086f802f9de6024298d00791539)
> 
> It seemed like allocating a work buffer once and then sharing it for all instances made the most sense. Alternatively it could be made per-instance at the expense of more heap. Or on demand with a bit more effort:
> 
> ```
> std::unique_ptr<uint8_t[]> work_buffer{new uint8_t[BSEC_MAX_WORKBUFFER_SIZE]};
> this->bsec_status_ = bsec_set_configuration(config.data(), BSEC_MAX_PROPERTY_BLOB_SIZE, reinterpret_cast<uint8_t *>(work_buffer.get()), BSEC_MAX_WORKBUFFER_SIZE);
> ```
> 
> Interested to hear your thoughts on approach.

Hmm.. Let me summarize:

**Single global work buffer**
Pros:
1. Less and simpler code
2. Less cycles spent allocating
3. Higher reliability (once the memory is allocated at setup time, we are guaranteed that the memory is there)

Cons:
1. Memory is reserved, so it's not available to other modules when the bme680_bsec module is not in execution

**Per-instance work buffers**
Pros: Exactly as the single global work buffer
Cons: Like the single global work buffer approach, with the amount of reserved memory multiplied by the number of instances

**On-demand work buffers**
Pros:
1. Memory usage of the entire system drops when the bme680_bsec component is not in use, and so there's more memory available to other modules (which can to their own allocations on the heap)

Cons:
1. Code usage and code complication is higher, as we need to allocate in every point where there's the need of a work buffer
2. Error handling would be more complicated (assuming that we don't have exceptions when an allocation fails: I need to experiment with that)
3. Increase in processor usage, as some cycles need to be spent on the allocation

I won't consider the per-instance work buffers approach, as it is equal to the single global work buffer but with more cons.
I think the processor cycles spent for the allocations should be negligible.

The ESP8266 has 80KiB of user memory ([source](https://en.wikipedia.org/wiki/ESP8266))
The work buffer is 2KiB.
So we're talking about reserving 2,5% of the total available user memory on an ESP8266.
ESP32 is OK, as it has 4x the RAM of an ESP8266 ([source](https://en.wikipedia.org/wiki/ESP32)).

Since we're talking about a work buffer, which is not needed anymore after the BSEC library function is returned, the logical approach would be to reclaim that memory, and not to keep it reserved undefinitely (as that's not a DMA buffer, or a portion of memory that contains useful data that we need to keep).

However, if we don't reserve it, we lose something on reliability: we introduce more points of failure, as the allocation may fail at runtime, maybe after the module is in use for several days or months, when the memory is more fragmented. The price of that reliability is a 2,5% increase in static memory usage on the ESP8266. It's a trade-off.

However, higher static memory usage may result in other modules to fail to operate. I'm not sure how many modules may be loaded simultaneously in a single ESP8266, and what is their memory usage (static+dynamic)

I estimated the size of a single bme680_bsec object (only counting data members) to be about 360 bytes, including members I added with this PR, so that's 0,5% of the total user memory per object. When (and if) multi-bus support will be added to the bme680_bsec module, one can use up to 8 I2C busses (16 pins) on a ESP8266, for a total of 16 objects. That will use 8% of the memory, plus the 2,5% needed for the global work buffer.
That's extreme, but at that point, the ESP8266 couldn't do much more (as the hardware pins are exhausted), so that may be ok (although we have WiFi, local web server, and maybe other modules which may be loaded without using hw pins).

My first reaction to the global work buffer was: _what a waste of memory! we can free that up, so other modules can use that_
However, thinking more about that, I am evenly worried about failed allocations in the long time. I personally would like my BME680 sensors as realiable as the Oregon Scientific thermometer they're going to replace.

These are my thoughts, sorry if they are so long 😅 

I am torn between these 2 approaches. What do you think?

----------

> I also switched to loading the config into vectors so that data ends up on the heap as well. As they're not massive I wonder if it would be faster with multiple sensors to allocate them them once per instance at setup time rather than on every re-init, but haven't benchmarked that as I only have a single sensor on my test rig.

I don't think this is needed at all. It all depends on the compiler, but:

- in the worst case you can have the compiler that still uses the stack to "implement" the #include, then you use the heap and copy data into it. You can prove this is the case by adding this part, and keeping the work buffer on the stack (I cannot do that as I don't own an ESP8266)

- In the best case, the compiler would include this data alongside with the code, and we would have zero stack usage (this data would be counted in the code usage). If I remember correctly, this array would be put in the `rodata` assembler section. Usually, `const` variables should be put in that section, but the arrays I declared are already `const`, so I think we're out of luck here. This approach is even advised [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/performance/ram-usage.html#reducing-static-memory-usage)

However, this is not a work buffer: it's a buffer that contains preinitialized data. I think we should continue to use that as-is on the stack (read: similar to the way it was before this PR), and instead focus on the movement of the work_buffer on the heap, given the fact that the work_buffer is 4x larger than the config array. (And that could even be a separate PR)

Reply:
I thought about the 3 scenarios, and I like better the first one with the single global work buffer.
I tried allocating a too-big memory chunk with a vector, and the ESP32 bootlooped. That would be the simulation of what would happen if, after running for some months, there would be memory fragmentation and a single buffer of 2048 bytes (BSEC_MAX_WORKBUFFER_SIZE) won't be available. I think we can live with the preallocated buffer, which is similar in reliability to the stack buffer we have now; if in the future someone wants more available memory while using the bme680_bsec module, we can think about that again (code usage would be higher too).
I'll update my code with a single global buffer soon!

Reply:
Reading over your thoughts, I would concur that favoring reliability (and simplicity) over the memory saving is the way to go. I think that making sure a simple setup works on an ESP8266 is sufficient as if anyone wants to run multiple sensors as well as the webserver etc. an ESP32 is the more logical choice.